README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
### Introduction

This package provides S3 classes ("uts\_vector" and "uts\_matrix") for multivariate unevenly spaced time series. The individual time series are [uts](https://github.com/andreas50/uts) objects.

R users familiar with vectors and matrices should find it straightforward to work with the corresponding time series objects, because the usage of many functions (such as \[, \[\[, length, etc.) is completely analogous.

### Installation

This package is not yet available on CRAN, but can be installled from GitHub, either using the `R` package [devtools](https://cran.r-project.org/web/packages/devtools/index.html):

``` r
devtools::install_github("andreas50/uts", build_vignettes=TRUE)
devtools::install_github("andreas50/utsMultivariate")
```

or using the `R` package [remotes](https://cran.r-project.org/web/packages/remotes/index.html):

``` r
remotes::install_github("andreas50/uts")
remotes::install_github("andreas50/utsMultivariate")
```

### Sample Code

First, a few examples how to use a time series vector.

``` r
# Get sample unevenly-spaced time series vector
x <- ex_uts_vector()
x
#>      Name Datapoints                   start                     end
#> 1  apples          6 2007-11-08 07:00:00 EST 2007-11-09 15:15:00 EST
#> 2 oranges          5 2007-11-08 08:30:00 EST 2007-11-09 10:21:00 EST
```

``` r
# Plot the time series vector
plot(x, type="o", cex.axis=0.8)
```

![](man/uts_plot-1.svg)

``` r
# Get first and last observation value(!) of each time series
first(x)
#>  apples oranges 
#> 48.3750 53.2125
last(x)
#>  apples oranges 
#>   47.35   52.25

# Get first and last observation time(!) of each time series
start(x)
#>                    apples                   oranges 
#> "2007-11-08 07:00:00 EST" "2007-11-08 08:30:00 EST"
end(x)
#>                    apples                   oranges 
#> "2007-11-09 15:15:00 EST" "2007-11-09 10:21:00 EST"

# Insert new observation
x[as.POSIXct("2007-11-10 10:00:00"), "apples"] <- 45

# Sample the time series at a specific time point, using one of several supported interpolation methods
sample_values(x, as.POSIXct("2007-11-10"), interpolation="linear")
#>   apples  oranges 
#> 46.25333 52.25000

# Get maximum observation value of each time series
max(x)
#>  apples oranges 
#>   48.50   53.35

# Time series arithmetic
(x*2 + 5)$apples$values
#> [1] 101.75 102.00 101.75  99.00 100.00  99.70  95.00

# Apply a function to the individual time series of a time series vector
sapply(x, length)
#>  apples oranges 
#>       7       5
sapply(x, range)
#>      apples oranges
#> [1,]   45.0   51.70
#> [2,]   48.5   53.35

# Convert the time series vector to data.frame, i.e. flatten the data
as.data.frame(x)
#>                   time apples oranges
#> 1  2007-11-08 07:00:00 48.375      NA
#> 2  2007-11-08 08:01:00 48.500      NA
#> 3  2007-11-08 08:30:00     NA 53.2125
#> 4  2007-11-08 09:31:00     NA 53.3500
#> 5  2007-11-08 13:15:00 48.375      NA
#> 6  2007-11-08 14:45:00     NA 53.2125
#> 7  2007-11-09 07:30:00 47.000      NA
#> 8  2007-11-09 08:51:00 47.500      NA
#> 9  2007-11-09 09:00:00     NA 51.7000
#> 10 2007-11-09 10:21:00     NA 52.2500
#> 11 2007-11-09 15:15:00 47.350      NA
#> 12 2007-11-10 10:00:00 45.000      NA

# Conversely, convert data in flat (long) format to a time series vector
values <- data.frame(apples=1:10, oranges=letters[1:10], stringsAsFactors=FALSE)
uts_vector_wide(values, times=as.POSIXct("2015-01-01") + ddays(1:10))
#>      Name Datapoints          start            end
#> 1  apples         10 2015-01-02 EST 2015-01-11 EST
#> 2 oranges         10 2015-01-02 EST 2015-01-11 EST

# Convert data in wide format to a time series vector
values <- c(5, 6, 6, 7)
times <- as.POSIXct("2016-01-01") + dhours(1:4)
names <- c("a", "b", "a", "a")
uts_vector_long(values, times, names)
#>   Name Datapoints                   start                     end
#> 1    a          3 2016-01-01 01:00:00 EST 2016-01-01 04:00:00 EST
#> 2    b          1 2016-01-01 02:00:00 EST 2016-01-01 02:00:00 EST
```

Most operatios for time series vectors also apply to a time series matrix, so only a few examples are given here.

``` r
# Create a time series matrix by recycling the elements of a time series vector
y <- uts_matrix(x, nrow=2, ncol=3)
rownames(y) <- c("foo", "bar")
colnames(y) <- c("red", "green", "blue")
y
#>     red    green  blue  
#> foo uts[7] uts[7] uts[7]
#> bar uts[5] uts[5] uts[5]

# Extract first two rows, and second column (coming soon)
#y[, 1:2, "green"]

# Transpose
t(y)
#>       foo    bar   
#> red   uts[7] uts[5]
#> green uts[7] uts[5]
#> blue  uts[7] uts[5]
```


ChangeLog
2018-06-09  Andreas Eckner  <andreas@eckner.com>

    * Initial package release

R/sampling.R
###################################################
# Sampling values from a multivariate time series #
###################################################

#' Sample Values
#' 
#' Sample each time series in a \code{"uts_vector"} at given sampling times and combine the sampled values into a \code{matrix} or \code{data.frame}.
#'
#' @return A \code{matrix} if all sampled observation values are of the same \code{\link{type}}, and a \code{data.frame} otherwise. However, when using argument \code{drop=TRUE} the result may be simplified further.
#' @param x a \code{"uts_vector"} object where each individual \code{"uts"} as atomic observation values.
#' @param time_points a strictly increasing sequence of \code{\link{POSIXct}} date-times.
#' @param drop logical. If \code{TRUE} the result is coerced to the lowest possible dimension. See the examples and \code{\link{drop}}.
#' @param \dots arguments passed to \code{\link[uts]{sample_values.uts}} for sampling from the individual time series.
#' 
#' @examples
#' times <- as.POSIXct(c("2007-11-09", "2007-11-10"))
#' 
#' # Sample with last-point and linear interpolation
#' sample_values(ex_uts_vector(), times)
#' sample_values(ex_uts_vector(), times, interpolation="linear")
#' 
#' # Sample with and without dropping of length-one dimensions
#' sample_values(ex_uts_vector(), times[1])
#' sample_values(ex_uts_vector(), times[1], drop=FALSE)
#' 
#' # Store sampled values in data.frame if of different type
#' x <- ex_uts()
#' y <- uts(letters[1:6], x$times)
#' utsv <- c(x, y)
#' sample_values(utsv, times)
#' 
#' # Error, because not all time series have atomic observation values
#' \dontrun{sample_values(ex_uts_vector2(), times)}
sample_values.uts_vector <- function(x, time_points, ..., drop=TRUE)
{ 
  # Argument checking
  is_atomic <- sapply(x, function(x) is.atomic(x$values))
  if (!all(is_atomic))
    stop("Not all time series have atomic observation values")
  
  # Sample each time series
  out <- lapply(x, sample_values, time_points, ...)
      
  # Combine results to matrix if sampled values are of same type, and data.frame otherwise
  types <- sapply(out, typeof)
  if (length(unique(types)) <= 1)
    out <- do.call(cbind, out)
  else
    out <- do.call(data.frame, out)
  colnames(out) <- names(x)
  
  # Drop length-one dimensions
  if (drop)
    out <- drop(out)
  out
}


#' Extract or Replace Parts of a uts_vector
#'
#' The accessor method (\code{"["}) extracts either (i) a sub-sampled time series vector, (ii) a subset of the time series vector, or (iii) both. The replacement method (\code{"[<-"}) inserts observation values at the provided observation times, replacing observations values for already existing observation times (if any).
#' 
#' @param x a \code{"uts_vector"} object.
#' @param i either a strictly increasing sequence of \code{\link{POSIXct}} date-times, or a \code{"uts"} or \code{"uts_vector"} with \code{\link{logical}} observation values.
#' @param j index specifying the time series to extract or replace. The index can be a \code{numeric} or \code{character} vector or empty (missing) or \code{NULL}. Numeric values are coerced to integer as by \code{\link{as.integer}} (and hence truncated towards zero). Character vectors will be matched to the \code{\link{names}} of the object.
#' @param drop logical. If \code{TRUE} the result is coerced to the lowest possible dimension.
#' @param \dots further arguments passed to \code{\link[uts]{sample_values.uts}}.
#' 
#' @examples
#' ##############################
#' # Subsampling and Subsetting #
#' ##############################
#' 
#' # Extract subset of time series vector
#' x <- ex_uts_vector()
#' x[, 1]
#' x[, "oranges", drop=FALSE]
#' x[, 2:1]
#' 
#' # Subsampling using a POSIXct vector
#' x <- ex_uts_vector()
#' times <- as.POSIXct(c("2007-11-08 11:01:00", "2007-11-09 15:16:00"))
#' x[times]
#' x[times, interpolation="linear"]
#' 
#' # Subsampling using a "uts" with logical observation values
#' x <- ex_uts_vector()
#' x[ex_uts() > 48]
#' 
#' # Subsampling and subsetting at the same time
#' # 1.) using separate subsampling and subsetting indicies
#' x[times, "oranges"]
#' x[times, c(1, 2, 1)]
#' # 2.) using "uts_vector" with logical observation values
#' x[x > 48]
#' 
#'
#' ###############################
#' # Replacements and Insertions #
#' ###############################
#' 
#' # Replace subset of time series vector with a single new "uts" time series
#' x <- ex_uts_vector()
#' x[, "oranges"] <- uts(values=50, times=Sys.time())
#' x[, "nuts"] <- head(ex_uts(), 2)
#' x[, "apples"] <- NULL
#' 
#' # Same, but use "uts_vector" for replacing
#' x <- c(ex_uts_vector(), nuts=ex_uts())
#' x[, 1:2] <- c(uts(), ex_uts())
#' x[, c("apples", "oranges")] <- NULL
#' 
#' # Replace/insert observations of single time series
#' # 1.) Insert single value into the first time series 
#' # 2.) Insert multiple values into the time series named "oranges"
#' x <- ex_uts_vector()
#' x[Sys.time(), 1] <- 50
#' x[Sys.time() + ddays(1:2), "oranges"] <- c(52, 53)
#' 
#' # Replace/insert observations of multiple time series
#' # 1.) Insert single value into every time series
#' # 2.) Insert the same multiple values into every time series
#' x <- ex_uts_vector()
#' x[Sys.time(), ] <- 51
#' x[Sys.time() + ddays(1:2), ] <- c(52, 53)
#' 
#' # Take replacement times from "uts" or "uts_vector" with logical observation values
#' x <- ex_uts_vector()
#' x[ex_uts() > 48] <- 5
#' x[x == 5] <- 6
`[.uts_vector` <- function(x, i, j, drop=TRUE, ...)
{
  # Extract subset time series vector
  if (!missing(j)) {
    x <- unclass(x)
    x <- x[j]
    x[!sapply(x, is.uts)] <- NULL
    
    if (length(x) == 0)
      return(uts_vector())
    else if ((length(x) == 1) && drop)
      x <- x[[1]]
    else
      class(x) <- class(uts_vector())
  }
  if (missing(i))
    return(x)
  
  # Check argument consistency
  num_ts <- ifelse(is.uts_vector(x), length(x), 1)
  num_selector_i <- ifelse(is.POSIXct(i) | is.uts(i), 1, length(i))
  if ((num_ts > 1) && (num_selector_i > 1) && (num_ts != num_selector_i))
    stop("The dimension of the 'uts_vector' the and sampling points `i` does not match")
  
  # Special case if dimension of 'x' was dropped
  if (is.uts(x))
    return(x[i, ...])
  
  # Sample each element of 'x'
  if (is.POSIXct(i) || is.uts(i))
    sapply(x, "[", i, ...)
  else {
    for (k in seq_len(num_ts))
      x[[k]] <- x[[k]][i[[k]], ...]
    x
  }
}


#' @rdname sub-.uts_vector
#' 
#' @param value a vector of observation values, a \code{"uts"}, a or \code{"uts_vector"}.
`[<-.uts_vector` <- function(x, i, j, ..., value)
{
  # Argument checking
  if (missing(j))
    j <- seq_along(x)
  
  # Case 1: replace subset time series vector with new time series
  if (missing(i)) {
    # need to remove class, because could not determine how to call [.default (if even possible)
    x <- unclass(x)
    if (is.uts(value))
      value <- uts_vector(value)
    x[j] <- value
    class(x) <- class(uts_vector())
    return(x)
  }
  
  # Case 2: uts_vector with logical observation values determines the insertion times
  if (is.uts_vector(i)) {
    if (length(i) != length(j))
      stop("The lengths of the uts_vectors 'x' and 'i' do not match")
    
    for (k in seq_along(x))
      x[[k]][i[[k]]] <- value
    return(x)
  }
  
  # Case 3: insert values into single time series
  if ((length(j) == 1) && is.vector(value)) {
    x[[j]][i] <- value
    return(x)
  }
  
  # Case 4: insert the same values into multiple time series
  if (is.vector(value)) {
    for (pos in j)
      x[[pos]][i] <- value
    return(x)
  }
  
  # If none of the cases applies
  stop("This combination of arguments is not supported")
}


